File: mobile/src/main/java/jp/satorufujiwara/kotlin/DaggerApp.java
package jp.satorufujiwara.kotlin;

import com.squareup.leakcanary.RefWatcher;
import com.squareup.okhttp.OkHttpClient;

import android.app.Application;
import android.content.Context;

import javax.inject.Inject;

public class DaggerApp extends Application {

    private AppComponent appComponent;

    @Inject
    AppLifecycleCallbacks appLifecycleCallbacks;

    @Inject
    RefWatcher refWatcher;

    @Inject
    OkHttpClient okHttpClient;

    @Override
    public void onCreate() {
        super.onCreate();
        appComponent = AppComponent.Initializer.init(this);
        appComponent.inject(this);
        AppGlideModule.registerComponents(this, okHttpClient);
        appLifecycleCallbacks.onCreate(this);
    }

    @Override
    public void onTerminate() {
        appLifecycleCallbacks.onTerminate(this);
        super.onTerminate();
    }

    public AppComponent appComponent() {
        return appComponent;
    }

    public RefWatcher refWatcher() {
        return refWatcher;
    }

    public static AppComponent appComponent(final Context context) {
        return ((DaggerApp) context.getApplicationContext()).appComponent();
    }

}

File: mobile/src/main/kotlin/jp/satorufujiwara/kotlin/DaggerApp.kt
package jp.satorufujiwara.kotlin

import android.app.Application
import com.squareup.leakcanary.RefWatcher
import com.squareup.okhttp.OkHttpClient
import javax.inject.Inject

public open class DaggerApp : Application() {

    val appComponent: AppComponent by lazy { AppComponent.Initializer.init(this) }
    @Inject lateinit val appLifecycleCallbacks: AppLifecycleCallbacks
    @Inject lateinit val refWatcher: RefWatcher
    @Inject lateinit val okHttpClient: OkHttpClient

    override fun onCreate() {
        super.onCreate()
        appComponent.inject(this)
        AppGlideModule.registerComponents(this, okHttpClient)
        appLifecycleCallbacks.onCreate(this)
    }

    override fun onTerminate() {
        appLifecycleCallbacks.onTerminate(this)
        super.onTerminate()
    }

    public fun refWatcher(): RefWatcher {
        return refWatcher
    }

}
